module/zfs/zvol.c

```diff
@@ -1685,10 +1696,11 @@ zvol_remove_minors_impl(const char *name)
 				continue;
 
 			zvol_remove(zv);
-			zvol_free(zv);
+			taskq_dispatch(system_taskq, zvol_free_impl, zv,
+			    TQ_SLEEP);
 		}
 	}
-
+	taskq_wait_outstanding(system_taskq, 0);
```

Comment:
Yeah, how about moving the 
zv->zv_disk->private_data = NULL;
out of the zvol_free_impl() to the zvol_remove_minors_impl() under the lock ?